### PR TITLE
fix(lib): ensure webpack inline loaders stay prefixed

### LINKED ISSUE #3314: Support webpack inline loaders for route components
### What problem does this feature solve?

As a nuxt module author I want to specify a webpack loader only for a given route component using [webpack inline loaders](https://webpack.js.org/concepts/loaders/#inline)

In order to apply custom transformations to files that can not be explicitly targeted via webpack config. (for example when the same file is loaded twice with different loaders).

### Problem:

The component path is made relative by [utils#relativeTo](https://github.com/nuxt/nuxt.js/blob/dev/lib/common/utils.js#L152-L171) which causes a component path like `my-epic-loader?foo=bar!./pages/epic.vue` to be prefixed like `../my-epic-loader?foo=bar!./pages/epic.vue` - webpack can not handle this.

### What does the proposed changes look like?

I'd propose to modify [relativeTo](https://github.com/nuxt/nuxt.js/blob/dev/lib/common/utils.js#L152-L171) to split the path by `!`, then make it relative and prepend the loader again.

I'd be happy to provide a PR

<div align="right"><sub><em>This feature request is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c7042">#c7042</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by Xiphe**
when making a component path relative

fix https://github.com/nuxt/nuxt.js/issues/3314

---

This is meant to support the discussion in https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c7042

 - ~~should we do this only if a `enableInlineLoaders` flag is true in nuxt config for backwards compatibility? (Is using `!` a common use-case nuxt supports?)~~
 - ~~The logic could also just be applied in [`router.js`](https://github.com/nuxt/nuxt.js/blob/dev/lib/app/router.js#L24) that might be less intrusive~~
 - [x] I will add tests once a solution is confirmed by a maintainer

### CODE REVIEW

File: lib/common/utils.js [hunk 1]
```diff
@@ -156,6 +156,16 @@ export const relativeTo = function relativeTo() {
   // Resolve path
   let _path = r(...args)
 
+  // Keep webpack inline loader intact
+  if (_path.indexOf('!') !== -1) {
+    const [loader, __path, ...rest] = path.split('!')
+    if (rest.length) {
+      throw new Error('Unexpected use of multiple "!" in path')
```
Comment (clarkdo) [inline lib/common/utils.js]: @Xiphe Thanks for your contribution.
I have a question: if use multiple inline loader such like `vue-loader!custom-loader?modules!./test.vue`, here may throw exception, so do you thinks we need to consider multiple loaders introduced by multiple modules?

File: lib/common/utils.js [hunk 2]
```diff
@@ -156,6 +156,16 @@ export const relativeTo = function relativeTo() {
   // Resolve path
   let _path = r(...args)
 
+  // Keep webpack inline loader intact
+  if (_path.indexOf('!') !== -1) {
+    const [loader, __path, ...rest] = path.split('!')
```
Comment (clarkdo) [inline lib/common/utils.js]: What if treat last item of splitted path as real path ?

```js
const [loader, ...rest] = path.split('!')
const __path = rest.pop()
```

Comment (Xiphe) [reply lib/common/utils.js]: Totally true. I somehow missed that someone could use multiple loaders and assumed that a second `!` would mean s.o. used it in a file/folder name.

**Comment by Xiphe** [comment]
Thanks for the feedback @clarkdo! Totally makes sense, I will update the PR and add tests

**Comment by Xiphe** [comment]
PTAL @clarkdo 

I realised that the loader logic needs to be before `let _path = r(...args)` since `r` does make the path absolute. Therefore I decided to only take loaders into account when they're part of the second argument of `relativeTo`. Otherwise we'd need to first join the path -> then slice the loaders -> make it absolute -> make it relative. 

I don't think that `relativeTo('/root/dir', 'loaderPath', 'moreLoaderPath!', 'otherPath')` is a realistic use case.

**Comment by Xiphe** [comment]
Ok, tests are green now. Took me a while to debug the windows tests 😕

**Review by clarkdo — APPROVED** [review]
LGTM

**Pull request merged**